File: android/src/main/java/com/lcorg/capacitor_secure_storage/SecureStoragePlugin.java
package com.lcorg.capacitor_secure_storage;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "SecureStoragePlugin")
public class SecureStoragePlugin extends Plugin {

  static final class ErrorCode {

    static final String INVALID_KEY = "INVALID_KEY";
    static final String INVALID_VALUE = "INVALID_VALUE";
    static final String ITEM_NOT_FOUND = "ITEM_NOT_FOUND";
    static final String STORAGE_ERROR = "STORAGE_ERROR";

    private ErrorCode() {}
  }

  static final class ErrorMessage {

    static final String INVALID_KEY = "Key must be a non-empty string";
    static final String INVALID_VALUE = "Value must be a string";
    static final String ITEM_NOT_FOUND = "Item with given key does not exist";
    static final String STORAGE_WRITE = "Unable to write to secure storage";
    static final String STORAGE_READ = "Unable to read from secure storage";
    static final String STORAGE_KEYS = "Unable to list secure storage keys";
    static final String STORAGE_REMOVE = "Unable to remove item from secure storage";
    static final String STORAGE_CLEAR = "Unable to clear secure storage";
    static final String STORAGE_NOT_INITIALIZED = "Secure storage is not initialized";

    private ErrorMessage() {}
  }

  static final class MethodName {

    static final String SET = "set";
    static final String GET = "get";
    static final String KEYS = "keys";
    static final String REMOVE = "remove";
    static final String CLEAR = "clear";

    private MethodName() {}
  }

  static final class OptionKey {

    static final String KEY = "key";
    static final String VALUE = "value";

    private OptionKey() {}
  }

  static final class PluginName {

    static final String VALUE = "SecureStoragePlugin";

    private PluginName() {}
  }

  private EncryptedStorageHelper encryptedStorageHelper;

  public SecureStoragePlugin() {}

  SecureStoragePlugin(EncryptedStorageHelper encryptedStorageHelper) {
    this.encryptedStorageHelper = encryptedStorageHelper;
  }

  @Override
  public void load() {
    super.load();
    this.encryptedStorageHelper = new EncryptedStorageHelper(getContext());
  }

  @PluginMethod
  public void set(PluginCall call) {
    String key = call.getString(OptionKey.KEY);
    String value = call.getString(OptionKey.VALUE);
    if (!isValidKey(key)) {
      call.reject(ErrorMessage.INVALID_KEY, ErrorCode.INVALID_KEY);
      return;
    }

    if (value == null) {
      call.reject(ErrorMessage.INVALID_VALUE, ErrorCode.INVALID_VALUE);
      return;
    }

    try {
      storage().setString(key, value);
      call.resolve();
    } catch (Exception exception) {
      rejectStorageError(call, ErrorMessage.STORAGE_WRITE, exception);
    }
  }

  @PluginMethod
  public void get(PluginCall call) {
    String key = call.getString(OptionKey.KEY);
    if (!isValidKey(key)) {
      call.reject(ErrorMessage.INVALID_KEY, ErrorCode.INVALID_KEY);
      return;
    }

    try {
      call.resolve(valueResult(readValue(key)));
    } catch (ItemNotFoundException exception) {
      call.reject(exception.getMessage(), ErrorCode.ITEM_NOT_FOUND);
    } catch (Exception exception) {
      rejectStorageError(call, ErrorMessage.STORAGE_READ, exception);
    }
  }

  @PluginMethod
  public void keys(PluginCall call) {
    try {
      call.resolve(valueResult(JSArray.from(storage().keys())));
    } catch (Exception exception) {
      rejectStorageError(call, ErrorMessage.STORAGE_KEYS, exception);
    }
  }

  @PluginMethod
  public void remove(PluginCall call) {
    String key = call.getString(OptionKey.KEY);
    if (!isValidKey(key)) {
      call.reject(ErrorMessage.INVALID_KEY, ErrorCode.INVALID_KEY);
      return;
    }

    try {
      requireExistingKey(key);
      storage().remove(key);
      call.resolve();
    } catch (ItemNotFoundException exception) {
      call.reject(exception.getMessage(), ErrorCode.ITEM_NOT_FOUND);
    } catch (Exception exception) {
      rejectStorageError(call, ErrorMessage.STORAGE_REMOVE, exception);
    }
  }

  @PluginMethod
  public void clear(PluginCall call) {
    try {
      storage().clear();
      call.resolve();
    } catch (Exception exception) {
      rejectStorageError(call, ErrorMessage.STORAGE_CLEAR, exception);
    }
  }

  private static boolean isValidKey(String key) {
    return key != null && !key.trim().isEmpty();
  }

  private EncryptedStorageHelper storage() {
    if (encryptedStorageHelper == null) {
      throw new IllegalStateException(ErrorMessage.STORAGE_NOT_INITIALIZED);
    }
    return encryptedStorageHelper;
  }

  private String readValue(String key) throws ItemNotFoundException {
    String value = storage().getString(key);
    if (value == null) {
      throw new ItemNotFoundException();
    }
    return value;
  }

  private void requireExistingKey(String key) throws ItemNotFoundException {
    if (!storage().contains(key)) {
      throw new ItemNotFoundException();
    }
  }

  private static JSObject valueResult(Object value) {
    JSObject ret = new JSObject();
    ret.put(OptionKey.VALUE, value);
    return ret;
  }

  private static void rejectStorageError(PluginCall call, String message, Exception exception) {
    call.reject(message, ErrorCode.STORAGE_ERROR, exception);
  }

  private static class ItemNotFoundException extends Exception {

    ItemNotFoundException() {
      super(ErrorMessage.ITEM_NOT_FOUND);
    }
  }
}

File: ios/Sources/SecureStoragePlugin/SecureStoragePlugin.swift
import Capacitor
import Foundation
import SimpleKeychain

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitor.ionicframework.com/docs/plugins/ios
 */
@objc(SecureStoragePlugin)
public final class SecureStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    enum ErrorCode {
        static let invalidKey = "INVALID_KEY"
        static let invalidValue = "INVALID_VALUE"
        static let itemNotFound = "ITEM_NOT_FOUND"
        static let storageError = "STORAGE_ERROR"
    }

    enum MethodName {
        static let set = "set"
        static let get = "get"
        static let keys = "keys"
        static let remove = "remove"
        static let clear = "clear"
    }

    enum OptionKey {
        static let key = "key"
        static let value = "value"
        static let group = "group"
        static let accessibility = "accessibility"
        static let isSynchronizable = "isSynchronizable"
    }

    enum Storage {
        static let service = "cap_sec"
    }

    private enum ErrorMessage {
        static let invalidKey = "Key must be a non-empty string"
        static let invalidValue = "Value must be a string"
        static let invalidAccessibility = "Accessibility must be a valid Keychain accessibility option"
        static let invalidGroup = "Group must be a string"
        static let invalidSynchronizable = "isSynchronizable must be a boolean"
        static let itemNotFound = "Item with given key does not exist"
        static let storageWrite = "Unable to write to secure storage"
        static let storageRead = "Unable to read from secure storage"
        static let storageKeys = "Unable to list secure storage keys"
        static let storageRemove = "Unable to remove item from secure storage"
        static let storageClear = "Unable to clear secure storage"
    }

    private struct KeychainOptions {
        let accessibility: Accessibility
        let group: String?
        let isSynchronizable: Bool
    }

    public let identifier = "SecureStoragePlugin"
    public let jsName = "SecureStoragePlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: MethodName.set, returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: MethodName.get, returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: MethodName.keys, returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: MethodName.remove, returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: MethodName.clear, returnType: CAPPluginReturnPromise),
    ]

    @objc
    func set(_ call: CAPPluginCall) {
        guard let key = self.requireKey(call) else {
            return
        }

        guard let value = call.getString(OptionKey.value) else {
            return call.reject(ErrorMessage.invalidValue, ErrorCode.invalidValue)
        }

        guard let keychainOptions = self.requireKeychainOptions(call) else {
            return
        }

        let simpleKeychain = self.makeKeychain(keychainOptions)

        do {
            try simpleKeychain.set(value, forKey: key)
            call.resolve()
        } catch {
            call.reject(ErrorMessage.storageWrite, ErrorCode.storageError, error)
        }
    }

    @objc
    func get(_ call: CAPPluginCall) {
        guard let key = self.requireKey(call) else {
            return
        }

        guard let keychainOptions = self.requireKeychainOptions(call) else {
            return
        }

        let simpleKeychain = self.makeKeychain(keychainOptions)

        do {
            guard try self.requireExistingItem(key, in: simpleKeychain, call) else {
                return
            }
            let value = try simpleKeychain.string(forKey: key)
            call.resolve([OptionKey.value: value])
        } catch {
            call.reject(ErrorMessage.storageRead, ErrorCode.storageError, error)
        }
    }

    @objc
    func keys(_ call: CAPPluginCall) {
        guard let keychainOptions = self.requireKeychainOptions(call) else {
            return
        }

        let simpleKeychain = self.makeKeychain(keychainOptions)

        do {
            let keys = try simpleKeychain.keys()
            call.resolve([
                OptionKey.value: Array(keys),
            ])
        } catch {
            call.reject(ErrorMessage.storageKeys, ErrorCode.storageError, error)
        }
    }

    @objc
    func remove(_ call: CAPPluginCall) {
        guard let key = self.requireKey(call) else {
            return
        }

        guard let keychainOptions = self.requireKeychainOptions(call) else {
            return
        }

        let simpleKeychain = self.makeKeychain(keychainOptions)

        do {
            guard try self.requireExistingItem(key, in: simpleKeychain, call) else {
                return
            }
            try simpleKeychain.deleteItem(forKey: key)
            call.resolve()
        } catch {
            call.reject(ErrorMessage.storageRemove, ErrorCode.storageError, error)
        }
    }

    @objc
    func clear(_ call: CAPPluginCall) {
        guard let keychainOptions = self.requireKeychainOptions(call) else {
            return
        }

        let simpleKeychain = self.makeKeychain(keychainOptions)

        do {
            try simpleKeychain.deleteAll()
            call.resolve()
        } catch {
            call.reject(ErrorMessage.storageClear, ErrorCode.storageError, error)
        }
    }

    private func makeKeychain(_ options: KeychainOptions) -> SimpleKeychain {
        return SimpleKeychain(
            service: Storage.service,
            accessGroup: options.group,
            accessibility: options.accessibility,
            synchronizable: options.isSynchronizable
        )
    }

    private func requireKeychainOptions(_ call: CAPPluginCall) -> KeychainOptions? {
        guard let accessibility = self.requireAccessibility(call) else {
            return nil
        }
        let group = call.getString(OptionKey.group)
        guard group != nil || call.options[OptionKey.group] == nil else {
            call.reject(ErrorMessage.invalidGroup, ErrorCode.invalidValue)
            return nil
        }
        guard let isSynchronizable = self.requireSynchronizable(call) else {
            return nil
        }

        return KeychainOptions(
            accessibility: accessibility,
            group: group,
            isSynchronizable: isSynchronizable
        )
    }

    private func requireSynchronizable(_ call: CAPPluginCall) -> Bool? {
        guard let rawValue = call.options[OptionKey.isSynchronizable] else {
            return false
        }
        guard let value = rawValue as? Bool else {
            call.reject(ErrorMessage.invalidSynchronizable, ErrorCode.invalidValue)
            return nil
        }
        return value
    }

    /// Rejects the call with `INVALID_KEY` and returns `nil` unless the call carries a
    /// non-empty (once trimmed) `key` option.
    private func requireKey(_ call: CAPPluginCall) -> String? {
        guard let key = call.getString(OptionKey.key), self.isValidKey(key) else {
            call.reject(ErrorMessage.invalidKey, ErrorCode.invalidKey)
            return nil
        }
        return key
    }

    /// A valid key must be non-nil and non-empty once surrounding whitespace is trimmed.
    private func isValidKey(_ key: String) -> Bool {
        return !key.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    /// Rejects the call with `ITEM_NOT_FOUND` and returns `false` unless `key` exists in `keychain`.
    private func requireExistingItem(
        _ key: String,
        in keychain: SimpleKeychain,
        _ call: CAPPluginCall
    ) throws -> Bool {
        guard try keychain.hasItem(forKey: key) else {
            call.reject(ErrorMessage.itemNotFound, ErrorCode.itemNotFound)
            return false
        }
        return true
    }

    private func requireAccessibility(_ call: CAPPluginCall) -> Accessibility? {
        guard let rawValue = call.options[OptionKey.accessibility] else {
            return Accessibility.afterFirstUnlockThisDeviceOnly
        }
        guard let value = rawValue as? String else {
            call.reject(ErrorMessage.invalidAccessibility, ErrorCode.invalidValue)
            return nil
        }

        switch value {
        case "whenUnlocked":
            return Accessibility.whenUnlocked
        case "whenUnlockedThisDeviceOnly":
            return Accessibility.whenUnlockedThisDeviceOnly
        case "whenPasscodeSetThisDeviceOnly":
            return Accessibility.whenPasscodeSetThisDeviceOnly
        case "afterFirstUnlock":
            return Accessibility.afterFirstUnlock
        case "afterFirstUnlockThisDeviceOnly":
            return Accessibility.afterFirstUnlockThisDeviceOnly
        default:
            call.reject(ErrorMessage.invalidAccessibility, ErrorCode.invalidValue)
            return nil
        }
    }
}
